x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorOnlyRegistry.java

```diff
@@ -7,98 +7,39 @@
-        "cluster:admin/shutdown/delete",
-        // Node removal prevalidation API
-        PrevalidateNodeRemovalAction.NAME,
-        // Desired Nodes API
-        DeleteDesiredNodesAction.NAME,
-        GetDesiredNodesAction.NAME,
-        UpdateDesiredNodesAction.NAME,
-        GetDesiredBalanceAction.NAME,
-        DeleteDesiredBalanceAction.NAME
-    );
-
-    private final ClusterSettings clusterSettings;
-
-    public OperatorOnlyRegistry(ClusterSettings clusterSettings) {
-        this.clusterSettings = clusterSettings;
-    }
+public interface OperatorOnlyRegistry {
 
     /**
      * Check whether the given action and request qualify as operator-only. The method returns
      * null if the action+request is NOT operator-only. Other it returns a violation object
      * that contains the message for details.
      */
-    public OperatorPrivilegesViolation check(String action, TransportRequest request) {
-        if (SIMPLE_ACTIONS.contains(action)) {
-            return () -> "action [" + action + "]";
-        } else if (ClusterUpdateSettingsAction.NAME.equals(action)) {
-            assert request instanceof ClusterUpdateSettingsRequest;
-            return checkClusterUpdateSettings((ClusterUpdateSettingsRequest) request);
-        } else {
-            return null;
-        }
-    }
+    OperatorPrivilegesViolation check(String action, TransportRequest request);
 
-    private OperatorPrivilegesViolation checkClusterUpdateSettings(ClusterUpdateSettingsRequest request) {
-        List<String> operatorOnlySettingKeys = Stream.concat(
-            request.transientSettings().keySet().stream(),
-            request.persistentSettings().keySet().stream()
-        ).filter(k -> {
-            final Setting<?> setting = clusterSettings.get(k);
-            return setting != null && setting.isOperatorOnly();
-        }).toList();
-        if (false == operatorOnlySettingKeys.isEmpty()) {
-            return () -> (operatorOnlySettingKeys.size() == 1 ? "setting" : "settings")
-                + " ["
-                + Strings.collectionToDelimitedString(operatorOnlySettingKeys, ",")
-                + "]";
-        } else {
-            return null;
-        }
-    }
+    /**
+     * Checks to see if a given {@link RestHandler} is subject to restrictions. This method may return a {@link RestResponse} if the
+     * request is restricted due to operator privileges. Null will be returned if there are no restrictions. Callers must short-circuit
+     * the request and respond back to the client with the given {@link RestResponse} without ever calling the {@link RestHandler}.
+     * If null is returned the caller must proceed as normal and call the {@link RestHandler}.
+     * @param restHandler The {@link RestHandler} to check for any restrictions
+     * @return null if no restrictions should be enforced, {@link RestResponse} if the request is restricted
+     */
+    RestResponse checkRestFull(RestHandler restHandler);
```

Comment:
I wonder if a rename here would be useful too?
`OperatorPrivilegesViolation checkRestRestriction(...)`
`OperatorPrivilegesViolation validateForRestRestriction(...)`

Reply:
> What are the advantages of returnning a RestResponse?

It allows custom error code and custom error messages as defined by operator privileges. For example, fully restricted, depending on context could return a 404 or a 403 and IMO it makes sense for the same thing (operator privs) to define the error code/message as well. For example, we know that in serverless we want to return a 404 and behave as though that API is not available at all, but (future) outside of serverless it would likely be a 403. I imagine there might be other case where, dependent on which exact endpoint is hit for which flavor of deployment we would want more control over the response.  If this only returns a violation then then one response code would need to work across all api's and flavors of deployments.  I hadn't considered auditing much yet but it appears we don't directly audit `OperatorPrivilegesViolation` and if we did it would not **require**  changes to this contract.

Reply:
ignore my previous comment.. i think i found a better way to represent the contract here.  working through the change now and will comment to the effect when the changes are made.

Reply:
the contract has been updated in the latest round of changes. fully restricted will now return a `OperatorPrivilegesViolation`